text/0000-impl-trait-type-alias.md

```diff
@@ -0,0 +1,624 @@
+```
+
+Finally, note that `impl Trait` `let` declarations hide the concrete
+types of local variables:
+
+```rust
+let displayable: impl Display = "Hello, world!";
+
+// We know `displayable` implements `Display`.
+println!("{}", displayable);
+
+// ERROR: no method `len` on `impl Display`
+// We don't know the concrete type of `displayable`,
+// so we don't know that it has a `len` method.
+assert_eq!(displayable.len(), 5);
+```
+
+At first glance, this behavior doesn't seem particularly useful.
+Indeed, `impl Trait` in `let` bindings exists mostly for consistency with
+`const`s and `static`s. However, it can be useful for documenting the
+specific ways in which a variable is used. It can also be used to provide
+better error messages for complex, nested types:
+
+```rust
+// Without `impl Trait`:
+let x = (0..100).map(|x| x * 3).filter(|x| x % 5 == 0);
+
+// ERROR: no method named `bogus_missing_method` found for type
+// `std::iter::Filter<std::iter::Map<std::ops::Range<{integer}>, [closure@src/main.rs:2:26: 2:35]>, [closure@src/main.rs:2:44: 2:58]>` in the current scope
+x.bogus_missing_method();
+
+// With `impl Trait`:
+let x: impl Iterator<Item = i32> = (0..100).map(|x| x * 3).filter(|x| x % 5);
+
+// ERROR: no method named `bogus_missing_method` found for type
+// `impl std::iter::Iterator` in the current scope
+x.bogus_missing_method();
+```
+
+## Guide: `impl Trait` Type Aliases
+[guide-aliases]: #guide-aliases
+
+`impl Trait` can also be used to create type aliases:
+
+```rust
+use std::fmt::Debug;
+
+type Foo = impl Debug;
+
+fn foo() -> Foo {
+    5i32
+}
+```
+
+`impl Trait` type aliases, just like regular type aliases, create
+synonyms for a type.
+In the example above, `Foo` is a synonym for `i32`.
+The difference between `impl Trait` type aliases and regular type aliases is
+that `impl Trait` type aliases hide their concrete type from other modules
+(but not submodules).
+Only the `impl Trait` signature is exposed:
+
+```rust
+use std::fmt::Debug;
+
+mod my_mod {
+  pub type Foo = impl Debug;
+
+  pub fn foo() -> Foo {
+      5i32
+  }
+
+  pub fn use_foo_inside_mod() -> Foo {
+      // Creates a variable `x` of type `i32`, which is equal to type `Foo`
+      let x: i32 = foo();
+      x + 5
+  }
+}
+
+fn use_foo_outside_mod() {
+    // Creates a variable `x` of type `Foo`, which is equal to type `impl Debug`
+    let x = my_mod::foo();
+
+    // Because we're outside `my_mod`, using a value of type `Foo` as anything
+    // other than `impl Debug` is an error:
+    let y: i32 = my_mod::foo(); // ERROR: expected type `i32`, found type `Foo`
+}
+```
+
+This makes it possible to write modules that hide their concrete types from the
+outside world, allowing them to change implementation details without affecting
+consumers of their API.
+
+Note that it is sometimes necessary to manually specify the concrete type of an
+`impl Trait`-aliased value, like in `let x: i32 = foo();` above.
```

Comment:
Just to confirm (I'm going over the comment thread and the RFC in parallel): The text here is meant to imply that the following alternative definitions for `fn use_foo_inside_mod` would both be rejected as erroneous by the compiler, right?

Erroneous version 1:
```rust
pub fn use_foo_inside_mod() -> Foo {
    let x: Foo = foo();
    //~^ ERROR: insufficient information to infer concrete implementation for type `Foo`
    //~^ NOTE: possible concrete type: `i32`; try `let x: I32 = foo();` instead?
    x + 5
}
```

Erroneous version 2:
```rust
pub fn use_foo_inside_mod() -> Foo {
    let x = foo();
    //~^ ERROR: insufficient information to infer type for `x`
    //~^ NOTE: possible concrete type: `i32`; try `let x: I32 = foo();` instead?
    x + 5
}
```

(The important bits are the errors in the above; the notes are just me guessing at ways we *might* try to make the compiler leverage information from elsewhere in the module to help the user along, but I don't know what @eddyb things about doing that even if only for help/notes.)

Reply:
@pnkfelix Correct. In order to use any properties of the type other than the `Debug` bound (in `type Foo = Debug;`), you'd need to specify `let x: i32 = foo();` or similar.

Reply:
Okay, after reading further I think that the [reference-level explanation](https://github.com/cramertj/rfcs/blob/impl-trait-alias/text/0000-impl-trait-type-alias.md#reference-impl-trait-type-aliases) addreses this question (and confirms my understanding)